pkg/dmesg/watcher.go

```diff
@@ -0,0 +1,176 @@
+package dmesg
+
+import (
+	"context"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/leptonai/gpud/log"
+	"github.com/leptonai/gpud/pkg/process"
+)
+
+var DefaultWatchCommands = []string{
+	"dmesg --decode --time-format=iso --nopager --buffer-size 163920 -w || true",
+
+	// run last commands as fallback, in case "dmesg -w" flag only works in some machines
+	"dmesg --decode --time-format=iso --nopager --buffer-size 163920 -W || true",
+}
+
+type LogLine struct {
+	Timestamp time.Time
+	Facility  string
+	Level     string
+	Content   string
+
+	// Error is set when a dmesg command fails.
+	Error string
+}
+
+type Watcher interface {
+	Watch() <-chan LogLine
+	Close()
+}
+
+func NewWatcher() (Watcher, error) {
+	return newWatcher(DefaultWatchCommands...)
+}
+
+func newWatcher(cmds ...string) (Watcher, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	ch, err := watch(ctx, cmds...)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+	return &watcher{ch: ch, cancel: cancel}, nil
+}
+
+type watcher struct {
+	ch     <-chan LogLine
+	cancel context.CancelFunc
+}
+
+func (w *watcher) Watch() <-chan LogLine {
+	return w.ch
+}
+
+func (w *watcher) Close() {
+	w.cancel()
+}
+
+func watch(
+	ctx context.Context,
+	cmds ...string,
+) (<-chan LogLine, error) {
+	ch := make(chan LogLine, 1000)
+	p, err := process.New(
+		process.WithCommand(cmds...),
+		process.WithRunAsBashScript(), // need to run as bash script when dmesg commands are complicated
+	)
+	if err != nil {
+		return nil, err
+	}
+	if err := p.Start(ctx); err != nil {
+		return nil, err
+	}
+	go read(ctx, p, ch)
+	go wait(ctx, p, ch)
+	return ch, nil
+}
+
+func wait(ctx context.Context, p process.Process, ch chan LogLine) {
+	defer func() {
+		// we don't close until either context is done or the process is exited
+		if err := p.Close(ctx); err != nil {
+			log.Logger.Warnw("failed to abort command", "err", err)
+		}
+		close(ch)
```

Comment:
Here has a race condition with read. Here close channel `ch` will cause the read panic when writing. So it's better to close ch in read goroutine.